### DISCUSSION TITLE
What's everybody using for joystick control

### DISCUSSION THREAD

**Discussion by UcfAerokid** [JSBSim-Team/jsbsim#572] [Q&A]
Greetings.  New to JSBsim, but not to simulation.  Is anybody using anything outside of FlightGear for joystick control?  I'm thinking of writing a UDP joystick driver to go straight to JSBSim, but if one already exists or there's a better alternative I'd like to try to use that instead, mapping joystick axes/buttons to FCS inputs.

I see old references to a stripchart utility that doesn't seem to exist any more anywhere.  Is anybody using anything more current for realtime stripcharting?

I wrote a DIS output interface using KDIS Open source library and I just got it working.  If anybody is interested in such a thing, I'd be interested in knowing the developer's approach to integrating other open source projects into jsbsim source.  I'll post something up in Show & Tell section when I have a bit of time.

Thanks in advance for the input.

UcfAerokid

**Comment by seanmcleod** [comment]
Over the years I've interfaced active inceptors like this one from Stirling Dynamics - https://www.stirling-dynamics.com/products/compact-stick and more recently from Brunner - https://www.brunner-innovation.swiss/product/cls-p-joystick/ with JSBSim. In my case I added code to JSBSim.

But you could look at using `FGUDPInputSocket` in order to update the relevant control properties from an external joystick process, local or remote.

In terms of real-time strip charting I've been using [PlotJuggler](https://www.plotjuggler.io/) for use with real recorded FTI, but have been meaning to add an interface to be able to connect JSBSim and PlotJuggler.

**Reply by UcfAerokid** [reply]
Thank you, Sean.  I was trying to avoid adding code to the codebase, so I started giving thought to writing a joystick->UDPInput driver.  I'll be using my own little DirectInput joystick class I wrote about 10 years ago, so it would be pretty specific and dated, but I'd share the rest of it to let everyone implement their own generic joystick code.

And thanks for sharing PlotJuggler.  Do you decode JSBSim UDP output in it?  Or repeatedly read the CSV as it goes?  Seems like UDP would be the most efficient, high-performance option.  I'll download and play with it to learn how to use it.

Many thanks again.

UcfAerokid

**Comment by seanmcleod** [comment]
Yep, so what I was suggesting is that your code making use of DirectInput takes the inceptor data and transmits it to JSBSim which has an existing `FGInputSocket` and `FGUDPInputSocket` class which listens for incoming network packets and reads them to then set property values, in your case properties to do with inceptor inputs.

In terms of PlotJuggler I've been using it with real FTI data that is recorded by an FTI system I developed. The data is loaded into PlotJuggler after flight, e.g. https://seanmcleod.github.io/2018/05/extra-300-spin/, so not real time streaming.

PlotJuggler supports a bunch of streaming protocols so my thinking was to add an output class to JSBSim that outputs in one of the supported streaming formats.

**Reply by UcfAerokid** [reply]
I'm going to look into a quick JSON output type to see if I can make a real-time driver for PlotJuggler.  That's a great suggestion.  Thank you.

**Comment by bcoconni** [comment]
FYI there is also [QtJSBSim](https://sourceforge.net/projects/qtjsbsim/) by @dpculp which interfaces joysticks with JSBSim via a socket.

**Comment by dpculp** [comment]
In QtJSBSim I use the SDLJoystick library to handle the joystick input.  QtJSBSim then communicates with an instance of the JSBSim standalone app over UDP sockets.

I would think building joystick handling inside JSBSim would be a bit out of JSBSim's scope.  On the other hand I could be wrong.

-- Dave

**Reply by UcfAerokid** [reply]
Thanks, Dave.  That's a great frontend.  Before I learned about it I was thinking of writing something similar in Wt for a browser-based GUI.  If I had the time I'd roll up a Windows build, but alas, my time is limited.  Thanks for introducing me to Cesium.

The code in setJSBSimControls() is pretty much what I had in mind for my approach to writing a joystick driver.

Thanks again.

**Reply by dpculp** [reply]
No problem.  One thing to watch out for with Cesium, if you decide to go that route, is that it requires the older QtWebKit module rather than the newer QtWebEngine module.  This is because the C++-to-Javascript bridge that I'm using to send data to the Cesium window doesn't exist in QtWebEngine.

-- Dave
